### ISSUE TITLE
User reports invalid signatures when constructing `ECDigitalSignature` from `ECDigitalSignature.fromRS()`

### ISSUE THREAD

**Issue by Christewart** [bitcoin-s/bitcoin-s#4309]
Start of the conversation here: https://suredbits.slack.com/archives/CAZSHL3EG/p1651611392673269?thread_ts=1651104885.625809&cid=CAZSHL3EG

From the user: 

>there is seems to be some issues with the way signatures are created from R,S coordinates. When my signer produces R.S coordinates with leading  0 bit the signature created with ECDigitalSignature.fromRS is accepted by bitcoind. But if my signer produces coordinate with leading 1, the 0 byte is correctly added but the signature is rejected. Please advise.

Here is an example of a transaction that gets rejected

>020000000253d48c625bf87d438c8449c926dc4f4fe13a76735aa6fabad0221caaf3393a9b010000006b48304502210093c84b241df86e5683f47bba94388363b7e85226baa818b9e32e4d42f5a0119502211c0ad68bff6631f192fb649fd2174dd5211322245f9a8d4ead13cf18c69747da012102181de645d5018fd757b99772bbd56745b32166582239f10fee52f1b2552c7818ffffffff2b801641bc503b10441e6638b70911a6ce88f0334852898e6302d0362a1ca0da000000006a473044022073216ad2cb2b3948fe4d8c8ce152e245d5d67df9669a326783be52308b7abf4f02202e9ea9a557829f03f3b65565458858674e73fff87c25a887cb6fbecc5b691d8a012102181de645d5018fd757b99772bbd56745b32166582239f10fee52f1b2552c7818ffffffff0210270000000000001976a91496bfaef29ba1e9979268ccd41724fc25b983b64588ac60176a00000000001976a914bc6050723955b77d7979bac75c66feb91dac6c1588ac00000000

Howeverhere is the same tx with a valid signature:

>020000000253d48c625bf87d438c8449c926dc4f4fe13a76735aa6fabad0221caaf3393a9b010000006a4730440220293825055be1e7f844e135bc4870273317ae676e950ea64ce459ec8b28be3229022055ed18c4e7166bd6c1ea84373bf4b3113017973c4a72af6019527c98b2e4b7af012102181de645d5018fd757b99772bbd56745b32166582239f10fee52f1b2552c7818ffffffff2b801641bc503b10441e6638b70911a6ce88f0334852898e6302d0362a1ca0da000000006a473044022056f2f3a95eba944ab49255f36f18f862c9ac0e4b2c01b4d8d86b035a49c1fe1502204c23a65486fa13361f5387cbe5359ef3d95bfaf2bc630d6d2c38d0eab4eb4954012102181de645d5018fd757b99772bbd56745b32166582239f10fee52f1b2552c7818ffffffff0210270000000000001976a91496bfaef29ba1e9979268ccd41724fc25b983b64588ac60176a00000000001976a914bc6050723955b77d7979bac75c66feb91dac6c1588ac00000000

Here is a branch I was using to debug: https://github.com/Christewart/bitcoin-s-core/tree/2022-05-03-debug-sig

This is the check specifically that seems to failing: https://github.com/Christewart/bitcoin-s-core/blob/ffcf2cd0fa83621d86843b1a5a0720acf363342f/crypto/src/main/scala/org/bitcoins/crypto/DERSignatureUtil.scala#L133

The user reports that these implementations work to produce a valid digital signature according to the bitcoin network, but these implementations fail our test suite

```scala
def fromRS(r: BigInt, s: BigInt): ECDigitalSignature = {
    val rHalf = ByteVector(0x02.toByte, ByteVector(r.toByteArray).size.toByte)
      .++(ByteVector(r.toByteArray))
    val sHalf = ByteVector(0x02.toByte, ByteVector(s.toByteArray).size.toByte)
      .++(ByteVector(s.toByteArray))
    val header = ByteVector(0x30.toByte, rHalf.length + sHalf.length)
    val bytes: ByteVector = {
      header.++(rHalf).++(sHalf).++(ByteVector(HashType.sigHashAll.byte))
    }
    val sig = DERSignatureUtil.lowS(ECDigitalSignature.fromBytes(bytes))
    assert(DERSignatureUtil.isValidSignatureEncoding(sig))
    sig
  }
```

or

```scala
def fromRS(r: BigInt, s: BigInt): ECDigitalSignature = {
    val rHalf = ByteVector(0x02.toByte, r.toByteArray.length)
      .++(ByteVector(r.toByteArray))
    val sHalf = ByteVector(0x02.toByte, s.toByteArray.length)
      .++(ByteVector(s.toByteArray))
    val header = ByteVector(0x30.toByte, rHalf.length + sHalf.length)
    val bytes: ByteVector = {
      header.++(rHalf).++(sHalf).++(ByteVector(HashType.sigHashAll.byte))
    }
    val sig = ECDigitalSignature.fromBytes(bytes)
    assert(DERSignatureUtil.isValidSignatureEncoding(sig))
    sig
  }
```

**Comment by benthecarman** [comment]
![image](https://user-images.githubusercontent.com/15256660/167056903-7053358e-ca3d-4af3-99ce-62622a5bb87b.png)

Looks like we aren't including the padding for the values. These are needed for high r and high s values.

https://b10c.me/blog/006-evolution-of-the-bitcoin-signature-length/

**Comment by nkohen** [comment]
I tested this and we do actually include the padding correctly. As far as I can tell the only issue here is that `DERSignatureUtil.isValidSignatureEncoding` assumes the presence of a SigHash byte (which is not true of `ECDigitalSignature` which can be constructed with or without one, and in particular the current `ECDigitalSignature.fromRS` method does not include a SigHash byte).

The two quicker solutions here seem to be either to add docs so that people know that if they use `fromRS` they have to add their own SigHash byte (probably would be nice to have a method for detecting and adding SigHash bytes) OR we can overload `fromRS` to also have a version which takes in and adds a SigHash. In either case docs should be added to `DERSignatureUtil.isValidSignatureEncoding` to state that the assumption is that there is a SigHash byte.

I personally think including all of these things would be nice to have, and will open a PR and see what people think.

**Comment by Christewart** [comment]
I definitely think we should improve the API. Maybe a `Option[HashType]` or something like that as a param. That will make the user of the API make the conscious decision if they want to include it or not.

**Comment by Christewart** [comment]
fixed in #4320
